### PR TITLE
"writeOnce"/"readOnly" flag for questions

### PR DESCRIPTION
**Pull request by Sergei-Poliakov**
## Context

[ATX-271](https://broadinstitute.atlassian.net/browse/ATX-271)
Implement "writeOnce" flag on the question-level.
For now, we decide to keep both features – activity-level & question-level.
Assumption: user update only fields, but they can't recall submission(like changing their mind 'Consent'->'Do Not Consent')

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [x] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java]:
```suggestion
                        "Could not find activity instance with guid=%s while getting question with id=%d and stableId=%s",
```

File: pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java]: Can we wrap in an `if/else` block so we check `isWriteOnce()` before making a database query?

**Review by yufengwng — COMMENTED** [review]
Looking good so far. Should we update `PatchFormAnswersRoute`? If the question is read-only, then we shouldn't allow updating answers right?

**Comment by bskinner** [comment]
I don't see any associated API documentation for these changes. There needs to be details on the new flags, and the documentation for the answer PATCH endpoint describing how its behavior is going to change.

File: pepper-apis/docs/specification/src/components/responses.yml
Comment (bskinner) [inline pepper-apis/docs/specification/src/components/responses.yml:79]: The question's readonly property also needs to be documented in the schemas for the Question objects.

File: pepper-apis/docs/specification/src/components/schemas.yml
Comment (bskinner) [inline pepper-apis/docs/specification/src/components/schemas.yml]: Given this is a non-nullable, optional property, this needs a default value defined. Since this also changes the default behavior when `true`, It would also more helpful to write the `description` from the perspective of "What changes when this is `true`?" instead of "What does `false` mean?"

```suggestion
    readonly:
      type: boolean
      description: if `true`, the client should not persist any changes to the question's answer. Any requests attempting to update this question's answer will result in an error.
      default: false
```

**Comment by yufengwng** [comment]
Looks like PR is in good shape. @ssettipalli or @MocanaAtBroad, what do you think?

**Comment by Sergei-Poliakov** [comment]
any updates here? can we merge it to develop?

**Comment by ssettipalli** [comment]
Sergei:
General team policy:
You don't need all the reviewers to review and approve a PR.
Once you have approval from at  least one reviewer you can go ahead and merge the PR.
If any of other reviewers reviewed the PR and had any concerns or changes requested, you need to address them.

Looks like Blake and Yufeng already reviewed this PR. Once you address and resolve their comments/concerns/changes requested, one of them will approve and you are good to merge the PR.

**Review by yufengwng — APPROVED** [review]
Looks good, don't see anything else that jumps out at me. Where can we see the Angular SDK changes? Is that coming in a different PR?

**Comment by Sergei-Poliakov** [comment]
> Looks good, don't see anything else that jumps out at me. Where can we see the Angular SDK changes? Is that coming in a different PR?

yup, currently blocked by this PR

**Pull request merged**
